File: Mapping_landscape_ABM/1_filter.R
require(tidyverse)
require(reticulate)
require(glmnet)
use_condaenv("conda_env", required = TRUE)


sbert = import("sentence_transformers")
model = sbert$SentenceTransformer("sentence-transformers/all-MiniLM-L6-v2")

# ---

labels = read_tsv("Mapping_landscape_ABM/Data/data_selected_labels_2.csv")

emb = model$encode(labels$abstract)

z = function(x) (x - mean(x))/sd(x)
for(i in 1:ncol(emb)) emb[,i] = z(emb[,i])


nrun = 100
lambdas = exp(seq(-10, 10, 1))
res = array(dim = c(nrun, length(lambdas), 2))
set.seed(42)
for(i in 1:nrun){
  print(i)
  for(j in 1:length(lambdas)){
    sel = sample(nrow(labels), round(nrow(labels)*.8))
    m = glmnet(emb[sel,], labels$out_of_scope[sel], family = "binomial", alpha = 0, lambda = lambdas[j])
    pred = predict(m, newx = emb[-sel,], type = "class") |> as.integer()
    res[i, j, 1] = mean(pred == labels$out_of_scope[-sel])
    pred = predict(m, newx = emb[-sel,], type = "response")
    res[i, j, 2] = -2*sum(log(pred) * labels$out_of_scope[-sel] + log(1-pred) * (1-labels$out_of_scope[-sel]))
    }
  }

colMeans(res[,,1]) |> plot()
colMeans(res[,,2]) |> plot()

# RERUN ----

best_lambda = lambdas[which.min(colMeans(res[,,2]))]
m = glmnet(emb, labels$out_of_scope, family = "binomial", alpha = 0, lambda = best_lambda)

# APPLY ----

data = read_csv("Mapping_landscape_ABM/Data/data_cleaned_2.csv")

data_emb = model$encode(data$Abstract_cleaned)
for(i in 1:ncol(data_emb)) data_emb[,i] = z(data_emb[,i])

label = predict(m, newx = data_emb, type = "class") |> as.integer()
label = predict(m, newx = data_emb, type = "response") 

data_filtered = data |> filter(label == 0)

write_csv(data_filtered, "Mapping_landscape_ABM/Data/data_cleaned_filtered_2.csv")

File: Mapping_landscape_ABM/1_filter.py
# libs
import numpy as np
import pandas as pd
from sentence_transformers import SentenceTransformer
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import log_loss, accuracy_score
from tqdm import trange


labels = pd.read_csv("data/filtering_labels.csv")

model = SentenceTransformer("sentence-transformers/all-MiniLM-L6-v2")
emb = model.encode(labels["abstract"].tolist(), show_progress_bar=True)

# z-score features (column-wise)
scaler = StandardScaler()
emb = scaler.fit_transform(emb)

y = labels["out_of_scope"].values

nrun = 100
lambdas = np.exp(np.arange(-10, 11))
res_acc = np.zeros((nrun, len(lambdas)))
res_ll  = np.zeros((nrun, len(lambdas)))

rng = np.random.default_rng(42)

for i in trange(nrun):
    idx = rng.permutation(len(y))
    split = int(len(y) * 0.8)
    train_idx, test_idx = idx[:split], idx[split:]

    Xtr, Xte = emb[train_idx], emb[test_idx]
    ytr, yte = y[train_idx], y[test_idx]

    for j, lam in enumerate(lambdas):
        clf = LogisticRegression(
            penalty="l2",
            C=1/lam,            
            solver="lbfgs",
            max_iter=5000
        )
        clf.fit(Xtr, ytr)

        y_pred = clf.predict(Xte)
        y_prob = clf.predict_proba(Xte)[:,1]

        res_acc[i, j] = accuracy_score(yte, y_pred)
        res_ll[i, j]  = log_loss(yte, y_prob)


mean_ll = res_ll.mean(axis=0)
best_lambda = lambdas[np.argmin(mean_ll)]

print("Best lambda:", best_lambda)



final_clf = LogisticRegression(
    penalty="l2",
    C=1/best_lambda,
    solver="lbfgs",
    max_iter=5000
)
final_clf.fit(emb, y)



data = pd.read_csv("1_data/data_cleaned.csv")

data_emb = model.encode(
    data["Abstract_cleaned"].tolist(),
    show_progress_bar=True
)
data_emb = scaler.transform(data_emb)

probs = final_clf.predict_proba(data_emb)[:,1]
data["out_of_scope_prob"] = probs



data_filtered = data[data["out_of_scope_prob"] < 0.5]
data_filtered.to_csv("1_data/data_cleaned_filtered.csv", index=False)
